src/Storages/RabbitMQ/StorageRabbitMQ.cpp

```diff
@@ -450,23 +458,52 @@ void StorageRabbitMQ::bindQueue(size_t queue_id)
 
     AMQP::Table queue_settings;
 
-    queue_settings["x-max-length"] = queue_size;
+    /// Check user-defined settings.
+    if (!queue_settings_list.empty())
+    {
+        for (const auto & setting : queue_settings_list)
+        {
+            Strings setting_values;
+            splitInto<'='>(setting_values, setting);
+            assert(setting_values.size() == 2);
+            String key = setting_values[0], value = setting_values[1];
+
+            std::unordered_set<String> integer_settings = {"x-max-length", "x-max-length-bytes", "x-message-ttl", "x-expires", "x-priority", "x-max-priority"};
+            std::unordered_set<String> string_settings = {"x-overflow", "x-dead-letter-exchange", "x-queue-type"};
+
+            if (integer_settings.find(key) != integer_settings.end())
+                queue_settings[key] = parse<uint64_t>(value);
+            else if (string_settings.find(key) != string_settings.end())
+                queue_settings[key] = value;
+            else
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Unsupported queue setting: {}", value);
+        }
+    }
 
-    if (!deadletter_exchange.empty())
+    /// Impose default settings if there are no user-defined settings.
+    if (!queue_settings.contains("x-max-length"))
+    {
+        queue_settings["x-max-length"] = queue_size;
+    }
+    if (!queue_settings.contains("x-dead-letter-exchange") && !deadletter_exchange.empty())
+    {
         queue_settings["x-dead-letter-exchange"] = deadletter_exchange;
-    else
+    }
+    else if (!queue_settings.contains("x-overflow"))
+    {
+        /// Define x-overflow only if there is not x-dead-letter-exchange, because it will overwrite the expected behaviour.
```

Comment:
But we will define x-overflow if queue_settings contains x-dead-letter-exchange. Or I missed something?

Reply:
Then probably it is intentional. Setting `x-dead-letter-exchange` has some cases, when it changes behaviour of rabbitmq queue and `x-overflow` is one of them, so it also makes sence to have `x-overflow` and `x-dead-letter-exchange` settings at the same time.
But by default, when there are no user-defined settings, there can be a separare setting in table settings `dead-letter-exchange` (it was added before the ability to add queue settings) so if it was added, there should be no overriding behaviour - that is what this comment is about.

Reply:
I am not sure that having two different ways to set `x-dead-letter-exchange` setting is good idea. What do you think?

Reply:
I agree, but I was not sure it is ok to delete settings which existed for a long time.. But I think I'll delete it, I doubt it was used.